aws-lc-rs/src/cipher/streaming.rs

```diff
@@ -0,0 +1,790 @@
+        let (written, remainder) = out_buffer.split_at_mut(written_len);
+        Self { written, remainder }
+    }
+}
+
+impl BufferUpdate<'_> {
+    /// Returns the slice from the buffer that was modified by the operation.
+    #[must_use]
+    pub fn written(&self) -> &[u8] {
+        self.written
+    }
+
+    /// Returns the slice of the buffer that was not modified by the operation.
+    #[must_use]
+    pub fn remainder(&self) -> &[u8] {
+        self.remainder
+    }
+
+    /// Returns a mutable slice of the buffer that was not modified by the operation.
+    #[must_use]
+    pub fn remainder_mut(&mut self) -> &mut [u8] {
+        self.remainder
+    }
+}
+
+impl StreamingEncryptingKey {
+    #[allow(clippy::needless_pass_by_value)]
+    fn new(
+        key: UnboundCipherKey,
+        mode: OperatingMode,
+        context: EncryptionContext,
+    ) -> Result<Self, Unspecified> {
+        let algorithm = key.algorithm();
+        let mut cipher_ctx = LcPtr::new(unsafe { EVP_CIPHER_CTX_new() })?;
+        let cipher = mode.evp_cipher(key.algorithm);
+        let key_bytes = key.key_bytes.as_ref();
+        debug_assert_eq!(
+            key_bytes.len(),
+            <usize>::try_from(unsafe { EVP_CIPHER_key_length(*cipher) }).unwrap()
+        );
+        let iv = <&[u8]>::try_from(&context)?;
+        debug_assert_eq!(
+            iv.len(),
+            <usize>::try_from(unsafe { EVP_CIPHER_iv_length(*cipher) }).unwrap()
+        );
+
+        // AWS-LC copies the key and iv values into the EVP_CIPHER_CTX, and thus can be dropped after this.
+        if 1 != unsafe {
+            EVP_EncryptInit_ex(
+                cipher_ctx.as_mut_ptr(),
+                *cipher,
+                null_mut(),
+                key_bytes.as_ptr(),
+                iv.as_ptr(),
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        Ok(Self {
+            algorithm,
+            mode,
+            cipher_ctx,
+            context,
+        })
+    }
+
+    /// Updates the internal state of the key with the provided ciphertext `input`,
+    /// potentially writing bytes of ciphertext to `output`.
+    ///
+    /// The number of bytes written to `output` can be up to `input.len()`
+    /// plus the block length of the algorithm (e.g., [`Algorithm::block_len`]).
+    ///
+    /// # Errors
+    /// * Returns an error if the `output` buffer is smaller than the length of
+    ///   the `input` plus the algorithm's block length (e.g. [`Algorithm::block_len`]) minus one.
+    /// * May return an error if the length of `input` plus the algorithm's block length is larger than `i32::MAX`.
+    pub fn update<'a>(
+        &mut self,
+        input: &[u8],
+        output: &'a mut [u8],
+    ) -> Result<BufferUpdate<'a>, Unspecified> {
+        let min_outsize = input
+            .len()
+            .checked_add(self.algorithm().block_len())
+            .ok_or(Unspecified)?
+            - 1;
```

Comment:
That might be true for most/all(?) cipher modes, but I believe the API contract/documentation doesn't provide any guarantee that it will necessarily be anything less than (`input_len` + `block_len` - 1). So I went with the most conservative restriction for it.